### DISCUSSION TITLE
Two-way audio and doorbell with 2N intercom systems

### DISCUSSION THREAD

**Discussion by longzheng** [koush/scrypted#1243] [Show and tell]
# Instructions to set up Scrypted with an 2N intercom (e.g. IP Solo, IP Vario) supporting two-way audio and doorbell functionality.

Make sure your 2N device is on the latest firmware. Earlier firmwares had issues with ONVIF not completely working and video feeds randomly freezing (required device restart). [Since version 2.35](https://www.2n.com/en-AU/newsroom/simple-and-understandable-the-new-2n-intercom-licensing/) some of the necessary features such as ~ONVIF~, ~automations~ and custom sounds no longer require an additional Gold license. ([Gold license is still required for ONVIF and automation](https://wiki.2n.com/hip/conf/latest/en/3-rozdily-mezi-modely-a-licencovani-funkci/3-2-licencovani-funkci))

## ONVIF two way-audio

Log in to your 2N device. First we need to configure some intercom streaming settings so that two-way audio works.

In the "Services", click "Streaming".
- Check "RTSP Server Enabled"
- Check "Audio Stream Enabled"
- Check "Video Stream Enabled"
- Create a user account by specifying a username and password

<img width="1164" alt="image" src="https://github.com/koush/scrypted/assets/484912/e5ebb446-e850-4660-8149-8616c660dd25">

Expand the "Transmission Quality Settings" section
- Check "UDP Unicast Enabled" (otherwise Scrypted won't be able to load the ONVIF stream)

Expand the "Fixed Streaming Profiles" section
- Change the "H.264 Video Parameters" settings to VGA/30fps/2048kbps (or your own preference)

<img width="1165" alt="image" src="https://github.com/koush/scrypted/assets/484912/aa4da0b5-35be-4d71-b784-8a87bc1b5837">

In Scrypted, install the `@scrypted/onvif` plugin if you don't already have it

<img width="1069" alt="image" src="https://github.com/koush/scrypted/assets/484912/a5d84c8b-a0e7-4a6b-99b2-683c7842fb60">

Open the "ONVIF Camera Plugin" and click "Add New" to add the 2N intercom. Fill in all the relevant details including the account username/password as configured in the 2N streaming settings.

<img width="717" alt="image" src="https://github.com/koush/scrypted/assets/484912/a3e04ec0-d3ff-4b39-a6b0-fb8c785c6b7a">

Once the device is created, make sure "Two Way Audio" is checked (it should be by default). Unfortunately this device doesn't support ONVIF doorbell events so we'll need to set up a manual webhook for doorbells later on.

<img width="666" alt="image" src="https://github.com/koush/scrypted/assets/484912/e4a5ca79-25e1-48cd-840c-ec892c4a370b">

Next we need to configure the streams. It's been observed that some 2N devices may put the MJPEG stream in `PROFILE1` and H.264 stream in `PROFILE2` which will cause issues when trying to use it with HomeKit. 

To figure out which stream we want, click "Streams" and then enable all 3 profiles as prebuffered streams and save.

![Screen Recording 2024-01-04 at 8 54 33 pm](https://github.com/koush/scrypted/assets/484912/b83511f1-6a4a-4e8a-a0fd-2f1afa50d54f)

Then click between all the different streams and find the one that has a "Detected Video/Audio Codecs" as `h264/pcm_mulaw` which is fully compatible with HomeKit without transcoding. 

It may be most likely `PROFILE1` or `PROFILE2` (you may also have multiple profiles, it doesn't matter we just need one).

<img width="661" alt="image" src="https://github.com/koush/scrypted/assets/484912/f4392f14-1eed-442e-a4ca-eea06d79c860">

After confirming which profile you need, go back to the "Manage" tab and deselect all the other profiles for "Prebuffered Streams" and set everything to your desired profile and save.

<img width="667" alt="image" src="https://github.com/koush/scrypted/assets/484912/2fd6513d-092d-4f65-8879-e1bb0b50b0e0">

Then switch back to your desired profile and change the "RTSP Parser" to "Scrypted (UDP)". Without changing this the audio stream won't work correctly.

<img width="665" alt="image" src="https://github.com/koush/scrypted/assets/484912/a19fd28f-4010-45bb-bc1b-8ce0f0664a80">

At this point the device should work perfectly as a camera with two-way audio. The next step is to make it a doorbell.

## Doorbell

We should first change the device type from "Camera" to "Doorbell" in the "Edit" tab and save. This will allow it to appear as a doorbell in HomeKit later.

<img width="668" alt="image" src="https://github.com/koush/scrypted/assets/484912/d7cd6992-a092-405a-abd3-4edceac87b1d">

Install the `@scrypted/webhook` and `@scrypted/dummy-switch` plugins if you don't already have it.

<img width="673" alt="image" src="https://github.com/koush/scrypted/assets/484912/eda29fc2-03d0-480f-9d3f-49bfc11b7f31">

<img width="664" alt="image" src="https://github.com/koush/scrypted/assets/484912/e91ec2df-14cc-45bf-afdb-64ef9c008e89">

Open the "Dummy Switch Plugin" and click "Add New" to create a new dummy switch. Name it appropriately.

<img width="664" alt="image" src="https://github.com/koush/scrypted/assets/484912/5f06742f-f9d4-494e-af67-ddfd5d90216a">

Change the "Reset Sensor" to "1" and save as we want this switch to behave like a button which gets toggled on and then off automatically from the button press on the 2N intercom.

<img width="669" alt="image" src="https://github.com/koush/scrypted/assets/484912/8f31dd7b-a561-4b58-81b7-fbcb57e60bd7">

Under "Extensions", enable the "Webhook" plugin so we can create a webhook that we can trigger from the 2N intercom. 

<img width="668" alt="image" src="https://github.com/koush/scrypted/assets/484912/c7d5a24f-802a-4b8d-a0fb-7b47a7a09286">

In the "Webhook" tab, click the "Create webhook" dropdown and select "OnOff" as we want to create a webhook that can turn on the switch. Click "Save".

<img width="665" alt="image" src="https://github.com/koush/scrypted/assets/484912/1ee228ed-c4c1-4fb4-8bd4-97ba45375be6">

To find the created webhook URL, click on the "Console" button 

<img width="663" alt="image" src="https://github.com/koush/scrypted/assets/484912/dfb173d4-b903-4cd7-9fc6-657483b32232">

In the console section that expands, make a note of the "Insecure Local Base URL". (You can copy this the whole output using the "Copy" button in the header, then paste into a text editor to copy the URL directly)

<img width="677" alt="image" src="https://github.com/koush/scrypted/assets/484912/5df204ef-eef4-4893-98f5-ec3e3223e511">

In my case it is `http://192.168.1.2:11080/endpoint/@scrypted/webhook/public/84/853dcf676198c598`. Because we'll want to trigger this "switch" to turn on from the button press, we will want to call `/turnOn`.

The combined URL becomes `http://192.168.1.2:11080/endpoint/@scrypted/webhook/public/84/853dcf676198c598/turnOn`. **Record this somewhere** since we'll need it later to configure in the 2N automation.

Then go back to the ONVIF device we previously set up, click on "Extensions" and enable "Custom Doorbell Button"

<img width="665" alt="image" src="https://github.com/koush/scrypted/assets/484912/bfa2ba48-ef3b-4bb0-b8cc-4adb5bbcfa3c">

Click on the "Custom Doorbell Button" tab and select the dummy switch we just created, and save.

<img width="665" alt="image" src="https://github.com/koush/scrypted/assets/484912/4ac2cf7e-0160-4d08-88d5-11d0762edc91">

Now we need to set up the 2N intercom to trigger the Scrypted switch via the webhook.

(Since we'll be using the call button to trigger the doorbell now, clear any existing SIP calling user you may have set up for the main unit button)

In the "Services" and "Automation" menu, click the "Edit" button on an function that's not being used. 

<img width="1118" alt="image" src="https://github.com/koush/scrypted/assets/484912/c5e58ec8-84e6-4085-afa4-0143deba07b4">

Drag the "[KeyPressed](https://wiki.2n.com/hip/auto/latest/en/3-udalosti-events#:~:text=switch%20(device%20opening)%3A-,KeyPressed,-The%C2%A0KeyPressed)" event (which is used to catch the button press event), "[SendHttpRequest](https://wiki.2n.com/hip/auto/latest/en/4-akce-action#:~:text=to%2010%20users.-,SendHttpRequest,-The%C2%A0SendHttpRequest)" action (to send the webhook call) and "[PlayUserSound](https://wiki.2n.com/hip/auto/latest/en/4-akce-action#:~:text=e.g.%20%22ext3%22.-,PlayUserSound,-The%C2%A0PlayUserSound)" action (to play a doorbell sound) options to the work area. Connect the "Event" from the "KeyPressed" to the "SendHttpRequest" and "PlayUserSound" events.

<img width="1267" alt="image" src="https://github.com/koush/scrypted/assets/484912/3b498569-ca88-4feb-9c01-3e97ef8ec90b">

In the "SendHttpRequest" box, paste the Scrypted webhook URL we previously saved (make sure it has the `/turnOn` at the end).

<img width="388" alt="image" src="https://github.com/koush/scrypted/assets/484912/ab5dc58b-9d91-407d-a37a-7a74611ee53c">

In the "PlayUserSound" box, set the sound you want to play when someone clicks on the button (because the traditional SIP calling sound won't be triggered). In my case I uploaded a custom user sound so I can enter in `1` for the first custom user sound. You can also set `*1` for a built-in ringtone. [See 2N documentation for more information](https://wiki.2n.com/hip/auto/latest/en/4-akce-action#:~:text=e.g.%20%22ext3%22.-,PlayUserSound,-The%C2%A0PlayUserSound).

Name the function and save it.

<img width="1470" alt="image" src="https://github.com/koush/scrypted/assets/484912/699ba23e-a518-44ed-8518-aab790cbb1e3">

Enable the function and save. Make sure it is "Running"

<img width="776" alt="image" src="https://github.com/koush/scrypted/assets/484912/049848eb-cea6-4ea5-af15-41020a4aaece">

Because we've disabled the SIP calling functionality, by default pressing the main unit button will trigger a dialling error sound (which will be played on top of the custom sound we defined in the automation). 

We can disable this in the "User Sounds" menu. Change the "Dialing Error Signaling" to "Silence" and save.

![image](https://github.com/koush/scrypted/assets/484912/b80732b9-c8ad-4004-b758-1bae0d5e2aee)

## Add to HomeKit

Now that everything is set up. Go back to Scrypted and the ONVIF camera device. In the "Extensions" tab, enable "HomeKit". Then click on the "HomeKit" tab, then "Pairing" and scan the QR code with your iOS device.

<img width="509" alt="image" src="https://github.com/koush/scrypted/assets/484912/106c3d42-2c52-498f-8843-0b5c64edd099">

If a QR code doesn't appear, the HomeKit plugin takes a while to initialize a new device so refresh after a few seconds.

Test two-way audio and pressing the 2N intercom should trigger a doorbell notification.

## Switch relay control

If you have a electromagnetic door/gate strike attached to the 2N intercom as a switch relay, you can use my [homebridge-2n-helios-switch Homebridge plugin](https://github.com/longzheng/homebridge-2n-helios-switch) to unlock/lock the switch in HomeKit.

**Comment by rohrsh** [comment]
Thank you so much for figuring all of this out. 

Small point for 2N IP Verso, which can have a full keypad installed. Enter "%1" as the KeyPressed event to action on the (first) doorbell button only. 

<img width="366" alt="image" src="https://github.com/koush/scrypted/assets/38928979/c3141976-a178-4474-bd98-4025582f0712">

**Comment by edelmaca** [comment]
If I change the IP Verso 2.0 to doorbell in scrypted Onvif plugin, the HomeKit extension says the camera can’t be exposed to HomeKit anymore. Is there a trick to do that anyway?

**Reply by longzheng** [reply]
Do you mean enabling this option?

![image](https://github.com/koush/scrypted/assets/484912/a9a2beb1-7485-415d-8c9c-214edefbb871)

I couldn't use that option because I don't think the 2N device has an ONVIF doorbell event name, so I don't have any experience with the issue you're describing.

**Reply by edelmaca** [reply]
No, the first step in the Doorbell part.
Edit: I mean the doorbell chapter of your howto.

**Reply by longzheng** [reply]
Oh if you've set up the camera already before in HomeKit, most likely you'll need to remove it and set it up again AFTER you've made all the changes.

If that still doesn't work you might need to reset pairing in the HomeKit options

<img width="1143" alt="image" src="https://github.com/koush/scrypted/assets/484912/b99d1cee-090f-462c-b5a8-0f6131a4cfd9">

**Reply by edelmaca** [reply]
Yep, that did it. I had to readopt it two times. The first time has failed as well. 
Removed it from HomeKit and reset the camera and the HomeKit plugin as well. After the second reset the 2N IP Verso 2.0 finally is working as it should.
If I had done you tutorial in one step and did not stop the first time before adding the doorbell webhook, it would have worked from the start.

**Comment by pgorod** [comment]
Hi, thanks for this write-up. I've gone through the steps but it's not exactly working; I need to ask a few questions to get some context, since I am new to Scrypted.

1. Does this require any of the paid stuff in Scrypted NVR, or can the 2N be used from Home Assistant without extra charges?
2. How do I get the camera into a card on the dashboard? Or is this meant to be used some other way?

Thanks in advance!

**Reply by longzheng** [reply]
1. It does not. I'm using it with just the free Scrypted installed onto a Windows computer.

2. I'm not using the Scrypted Home Assistant integration so I can't help with that.

**Comment by AboveNBeyondCode** [comment]
Thx for pointing this out, but I am not able to activate the video features without a gold lic. Paid Licenses: "Enhanced Video NO" using Firmware Version 2.45.0.59.2 on a 2N IP Verso. The only video option I can enable ist jpeg / mjpeg buy using: http:/IP//api/camera/snapshot?width=1280&height=960&fps=15 for a "video".  Is there a solution for this or do I have to buy the gold lic?

**Reply by longzheng** [reply]
Sorry you're right, RTSP and ONVIF is not included in the free license and requires Gold. https://wiki.2n.com/hip/conf/latest/en/3-rozdily-mezi-modely-a-licencovani-funkci/3-2-licencovani-funkci

**Comment by moritzj29** [comment]
Thanks so much for this detailed How-To! I just setup my new Verso 2.0 with scrypted and HKSV is working very nicely already!

The scrypted interface and the plugin configuration appears to have slightly changed over time, so let me contribute to this guide with a couple of remarks (based on scrypted v0.141.0):

### ONVIF plugin
When setting up the ONVIF camera in scrypted **un**check "automatically configure settings" -> this is not supported by 2N/axis (https://github.com/koush/scrypted/issues/1585).
The different streams are setup in the Verso configuration directly. The stream configuration is read by the ONVIF plugin, therefore setup the streams (resolution etc.) before adding the camera to scrypted.

### Setup snapshot URL
I found that the snapshot URL needs to be manually set in scrypted. Otherwise no preview is shown in scrypted or in the Home app until you actually start streaming.

First, check which authentication settings are in place for the Camera API:
<img width="998" height="690" alt="image" src="https://github.com/user-attachments/assets/da690186-5a62-4886-b795-1ac047db2bb3" />
I just left the default "Secure (TLS)" and "Digest" Authentication. You can also disable authentication here for the Camera API and skip the next step.
Create and enable a user account with the **same user and password as in the RTSP/ONVIF configuration**. Scrypted is using the same credentials for RTSP stream and JPEG.
<img width="1996" height="1694" alt="image" src="https://github.com/user-attachments/assets/ff239186-95f6-4297-b60f-780baf7ed1da" />
In the scrypted configuration set the correct Snapshot URL:
<img width="1406" height="1098" alt="image" src="https://github.com/user-attachments/assets/b101585e-6fd7-4ea8-85b4-4996efcac95a" />
Depending if you chose "Secure (TLS)" previously or not, you need to prefix the URL with HTTPS or plain HTTP. There is no certificate validation in place, it is working with Verso's self signed certificate without issue.

URL Format: `https://intercom_ip_address/api/camera/snapshot?width=W&height=H` (described in JPEG section [here](https://wiki.2n.com/hip/conf/latest/en/5-konfigurace-interkomu/5-4-sluzby/5-4-2-streamovani))
